### PR TITLE
Add internal determinism flag configuration for ORTModule

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Current implementation does not support configuration of determinism on ORT

This PR sets ORTModule's determinism based on torch's determinism flag. It also exposes two internal APIs to get/set determinism: `use_deterministic_algorithms` and `are_deterministic_algorithms_enabled`, which inherits the same names from PyTorch

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 1]
```diff
@@ -232,6 +237,12 @@ def _build_graph(self):
 
     def _get_session_config(self):
         """Creates and returns the session configuration to be used for the ExecutionAgent"""
+
+        if torch.are_deterministic_algorithms_enabled():
+            if self._debug_options.logging.log_level <= _logger.LogLevel.INFO:
+                warnings.warn("ORTModule's determinism will be enabled because PyTorch's determinism is enabled.",
```
Comment (suffiank) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:243]: Like it 👍

**Review by suffiank — DISMISSED** [review]

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 1 above)
Comment (suffiank) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:243]: Actually, did you mean to write >= INFO? https://docs.python.org/3/library/logging.html#levels

**Comment by suffiank** [comment]
Actually, I think you may have had the log level check right the 1st time. https://docs.python.org/3/library/logging.html#levels The lower the level the more output expected. Sorry, I may have mislead with earlier comment.

File: orttraining/orttraining/python/training/ortmodule/__init__.py [hunk 2]
```diff
@@ -74,6 +76,17 @@ def override_torch_cuda_manual_seed(seed):
 torch_cuda_manual_seed = torch.cuda.manual_seed
 torch.cuda.manual_seed = override_torch_cuda_manual_seed
 
+
+def use_deterministic_algorithms(enabled):
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/__init__.py:80]: Shall we only keep this internal for now and not allow users to enable/disable this for the time being (until we feel the need to explicitly allow users to change this).

File: orttraining/orttraining/python/training/ortmodule/_inference_manager.py [hunk 3]
```diff
@@ -93,21 +98,24 @@ def forward(self, *inputs, **kwargs):
             # If creating the execution agent for the first time, this skip check will not take effect.
             # It will only take effect on subsequent forward calls.
             create_execution_session = False
-            if self._skip_check.is_set(_SkipCheck.SKIP_CHECK_EXECUTION_AGENT) == False or \
-                not self._execution_agent:
+            if self._skip_check.is_set(_SkipCheck.SKIP_CHECK_EXECUTION_AGENT) is False or \
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_inference_manager.py:101]: Should we not continue using `==` and `!=` for values? I think for bools it does not matter. But to keep things consistent, we could probably use `==` and `!=` to denote equality of value and `is` to denote equality of identity?

File: orttraining/orttraining/python/training/ortmodule/_inference_manager.py [hunk 4]
```diff
@@ -93,21 +98,24 @@ def forward(self, *inputs, **kwargs):
...
+                    not self._execution_agent:
                 module_device = _utils.get_device_from_module(
                     self._original_module)
-                # The inference session should be created every time
-                # the graph was built or if the device changed between calls to forward
-                create_execution_session = build_graph or self._device != module_device
+
+                create_execution_session = (build_graph or self._device != module_device or
+                                            torch.are_deterministic_algorithms_enabled() is not
+                                            are_deterministic_algorithms_enabled())
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_inference_manager.py:108]: Do you think instead of keeping a global variable for determinism which could easily go out of consistency with what the session was created with, we could persist the `session_options` or more specifically `use_deterministic_compute` to compare `torch.are_deterministic_algorithms_enabled()` against?

File: orttraining/orttraining/python/training/ortmodule/__init__.py (hunk 2 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule/__init__.py:80]: onnxruntime repo is internal :) only torch_ort is public

File: orttraining/orttraining/python/training/ortmodule/_inference_manager.py (hunk 4 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule/_inference_manager.py:108]: SessionOptions is a per model instance configuration, whereas determinism is a framework configuration, which is the same behavior as for PyTorch `are_deterministic_algorithms_enabled`

File: orttraining/orttraining/python/training/ortmodule/_inference_manager.py (hunk 3 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule/_inference_manager.py:101]: Linter complained. Bools must be compared as `if cond` or `if var is True|False` as per PEP

File: orttraining/orttraining/python/training/ortmodule/__init__.py [hunk 5]
```diff
@@ -74,6 +76,17 @@ def override_torch_cuda_manual_seed(seed):
...
+    global ORTMODULE_IS_DETERMINISTIC
+    ORTMODULE_IS_DETERMINISTIC = enabled
+
+
+def are_deterministic_algorithms_enabled():
```
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule/__init__.py:85]: Defining APIs at `__init__.py` is a common practice https://github.com/pytorch/pytorch/blob/master/torch/__init__.py#L487

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 6]
```diff
@@ -232,6 +237,12 @@ def _build_graph(self):
 
     def _get_session_config(self):
         """Creates and returns the session configuration to be used for the ExecutionAgent"""
+
+        if torch.are_deterministic_algorithms_enabled():
```
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:241]: Both APIs always return the same value as we keep it in sync, but I can update to use ORT's

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py [hunk 7]
```diff
@@ -160,7 +160,7 @@ def _post_process_after_export(exported_model, enable_custom_autograd_function,
             is_pythonop_needed = True
             break
 
-    if is_pythonop_needed and log_level <= _logger.LogLevel.WARNING:
+    if is_pythonop_needed and log_level >= _logger.LogLevel.WARNING:
```
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:163]: As Suffian pointed out in the doc, `NOTSET < DEBUG < INFO < WARNING < ERROR < CRITICAL`.

When we configure log level to CRITICAL, only CRITICAL logs are shown, but when we configure it to WARNING, for instance, WARNING, ERROR and CRITICAL are shown. If INFO is set, INFO, WARNING, ERROR and CRITICAL are shown

We should read this if as `if log level is at least WARNING, then show it to the user`. Does it make sense

File: orttraining/orttraining/python/training/ortmodule/__init__.py [hunk 8]
```diff
@@ -88,5 +94,5 @@ def are_deterministic_algorithms_enabled():
 
 
 # ORTModule must be loaded only after all validation passes
-from .ortmodule import ORTModule
-from .debug_options import DebugOptions, LogLevel
+from .ortmodule import ORTModule # noqa: E402 E261
+from .debug_options import DebugOptions, LogLevel # noqa: E402 E261
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/__init__.py:98]: unintentional comments.

File: orttraining/orttraining/python/training/ortmodule/__init__.py (hunk 2 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/__init__.py:86]: My concern (for the time being) is not with the API but with how we are currently using it. We create a new execution agent whenever `torch.are_deterministic_algorithms_enabled()` is different from `ortmodule.are_deterministic_algorithms_enabled()`. With this being exposed, users may start using this API and unknowingly (by not setting the same on torch) trigger re-creation of the execution agent (since our logic is checking this value againts torch's).

Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/__init__.py:86]: Shall we instead just simply use `torch.are_deterministic_algorithms_enabled()` all the time instead of defining our own `are_deterministic_algorithms_enabled()`? What advantage does this bring us at this time to expose this if we anyway only reading the torch's implementation to set our determinism.

File: orttraining/orttraining/python/training/ortmodule/_inference_manager.py (hunk 4 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/_inference_manager.py:108]: I am ok with it being a framework level setting but just have concerns around it going out of sync with the torch's value and then causing unexpected results for the user.

File: orttraining/orttraining/python/training/ortmodule/__init__.py (hunk 8 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule/__init__.py:98]: Actually they are intended. They disable linter error for imports not being at the top of the file

File: orttraining/orttraining/python/training/ortmodule/_inference_manager.py (hunk 4 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule/_inference_manager.py:108]: Got it, but they are part of the `create_execution_session` check for each `forward` call. When they mismatch, we update it internally and rebuild the execution session to reflect the change

File: orttraining/orttraining/python/training/ortmodule/__init__.py (hunk 2 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule/__init__.py:86]: We keep our own to detect changes to determinism state. From one forward call to the next, if we only use pytorch's, how can we know pytorch change from true to false and otherwise? The idea is that our flag is set during initialization and then always compared to pytorch's for equality. Any mismatch leads to recreation of execution session. Makes sense?

Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule/__init__.py:86]: For your previous concern, I think you are right. If we expose it to the public `torch_ort` API, then the user could accidentally cause the execution session creation. I will append `_` to them until we decide about its utility as standalone APIs

Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/__init__.py]: Yeah, we need a way to determine change in pytorch determinism. And my thought was to persist the setting that we used to save the `SessionOptions` determinism and then compare against that so we don't need to keep a gloabl variable which could go out of sync (particularly if we always want to follow pytorch's determinism).
Let's go ahead and merge this now. And we can revisit this later if we feel that this causes confusion (or consistency issues).

**Review by baijumeswani — APPROVED** [review]
